File: Trees/Binary Tree/Python/binaryTree.py
from node import Node
import random as rd

#BST
class BinaryTree(Node): 
    def __init__(self):
        self.__root = None


    def insert(self, value):
        if (self.__root == None):
            self.__root = Node(value=value, parent=None)
            return
        pointer = self.__root
        while True:
            if value < pointer.value and pointer.left:
                pointer = pointer.left
            elif value >= pointer.value and pointer.right:
                pointer = pointer.right
            else: break

        if pointer.value > value: pointer.left = Node(value, pointer)
        else: pointer.right = Node(value, pointer)


    def remove(self, value):
        def min(pointer): 
            while pointer.left: # 
                pointer = pointer.left
            return pointer
        """
            while pointer.left: # 
                pointer = pointer.left
            return pointer
        """

        pointer = self.binarySearch(value)

        if not pointer:
            return
        
        parent = pointer.parent
        if parent:
            if parent.left and parent.left == pointer: atr = "left"
            else: atr = "right"

        """
        if pointer.left and pointer.right:
            ptr = min(pointer.right)
            pointer.value = ptr.value
            if ptr == ptr.parent.left:
                ptr.parent.left = None
            else: ptr.parent.right = None
        """
        # both childs
        if pointer.left and pointer.right:
            ptr = min(pointer.right)
            pointer.value = ptr.value
            if ptr == ptr.parent.left:
                ptr.parent.left = None 
            else: ptr.parent.right = None
        # one child
        elif pointer.left:    
            if pointer == self.__root:
                self.__root = pointer.left
            else:
                setattr(parent, atr, pointer.left)
                pointer.left.parent = parent
        elif pointer.right:
            if pointer == self.__root:
                self.__root = pointer.right
            else:
                setattr(parent, atr, pointer.right)
                pointer.right.parent = parent
        # leaf node 
        else: setattr(parent, atr, None)


    def rotateLR():
        pass
    
    
    def rotateLL():
        pass
    
    
    def rotateRR():
        pass
    
    
    def rotateRL():
        pass


    def balance():
        pass

    def pre_order_traverse_search(self, value, pointer=False): 
        if pointer == False: pointer = self.__root
        if (pointer == None): return
        
        if pointer.value == value:
            return pointer
        # Look for right child
        buffer = self.pre_order_traverse_search(value, pointer.right)
        if buffer != None: return buffer
        # Look for left child
        buffer = self.pre_order_traverse_search(value, pointer.left)
        if buffer != None: return buffer
    
    
    def post_order_traverse_search(self, value, pointer=False): 
        if pointer == False: pointer = self.__root
        if (pointer == None): return
        
        # Look for right child
        buffer = self.post_order_traverse_search(value, pointer.right)
        if buffer != None: return buffer
        # Look for left child
        buffer = self.post_order_traverse_search(value, pointer.left)
        if buffer != None: return buffer

        if pointer.value == value:
            return pointer
    
    
    def in_order_traverse_search(self, value, pointer=False): 
        if pointer == False: pointer = self.__root
        if (pointer == None): return
        
        # Look for right child
        buffer = self.in_order_traverse_search(value, pointer.right)
        if buffer != None: return buffer

        if pointer.value == value:
            return pointer

        # Look for left child
        buffer = self.in_order_traverse_search(value, pointer.left)
        if buffer != None: return buffer


    def binarySearch(self, value, pointer=False):
        if pointer==False:
            pointer = self.__root
        if pointer.value == value:
            return self.__root
        while True:
            if pointer.value>value and pointer.left!=None:
                pointer = pointer.left
            elif pointer.value<value and pointer.right != None:
                pointer = pointer.right
            else: 
                break
        if pointer.value == value:
            return pointer


    # Turn tree to array by pre order traversal
    def pre_order_to_array(self, pointer=False):
        if pointer == False: pointer = self.__root
        out = []
        
        out.append(pointer.value)        
        if pointer.left: out += self.pre_order_to_array(pointer.left) 
        if pointer.right: out += self.pre_order_to_array(pointer.right) 

        return out  


    # Turn tree to array by in order traversal
    def in_order_to_array(self, pointer=False):
        if pointer == False: pointer = self.__root
        out = []
        
        if pointer.left: out += self.in_order_to_array(pointer.left) 
        out.append(pointer.value)        
        if pointer.right: out += self.in_order_to_array(pointer.right) 

        return out  


    # Turn tree to array by post order traversal
    def post_order_to_array(self, pointer=False):
        if pointer == False: pointer = self.__root
        out = []
        
        if pointer.left: out += self.post_order_to_array(pointer.left) 
        if pointer.right: out += self.post_order_to_array(pointer.right) 
        out.append(pointer.value)        

        return out  


    def __add__(self, lst):
        for i in lst:
            self.insert(i)


    def __str__(self):
        return self.__repr__()


    def __eq__(self, __o: object) -> bool:
        return self.in_order_to_array() == __o.in_order_to_array()


    def __len__(self):
        return len(self.in_order_to_array())


    def height(self, pointer=False, height=False):
        if pointer == False: pointer = self.__root
        elif not pointer: return height
        if height == False: height = 0
        left = self.height(pointer.left, height+1)
        right = self.height(pointer.right, height+1)
        return max(left, right)


    # To clearly see which number to remove
    def display_highlight(self, to_remove=None, pointer=False, space=0, level=4):
        if pointer == False: pointer = self.__root
        if (pointer == None): return
        space += level
        self.display_highlight(to_remove, pointer.right, space)
        for i in range(level, space): print(end = '\033[92m' + ".")  
        if pointer.value != to_remove:
            print('\033[93m' + "[" + '\033[0m' + str(pointer.value) + '\033[93m' + "]<")
        else: 
            print('\033[93m' + "[" + '\033[91m' + str(pointer.value) + '\033[93m' + "]<")
        self.display_highlight(to_remove, pointer.left, space)
        print("" + '\033[0m', end="")


    # In Order Traversal Print 
    def display(self, pointer=False, space=0, level=4):
        if pointer == False: pointer = self.__root
        if not pointer: return
        space += level
        self.display(pointer.right, space)

        for i in range(level, space): print(end = ".")  
        print("[" + str(pointer.value) + "]<")
        self.display(pointer.left, space)
        print("", end="")


    def __repr__(self, pointer=False):
        if pointer == False: pointer = self.__root
        out = ""
        
        if pointer.left: out += self.__repr__(pointer.left) 
        out += str(pointer.value) + " "        
        if pointer.right: out += self.__repr__(pointer.right) 

        return out  


if __name__ == "__main__":
    tree = BinaryTree()
    tree + [30, 40, 10, 5, 15, 14, 25, 20]
    tree.display()
    print("____")
    tree.display()
    print(tree.height())
    print(tree.in_order_to_array())

File: 8_Trees/BST/JS/binary tree.js
class Node {
    constructor (value, parent) {
        this.parent = parent;
        this.value = value;
        this.left;
        this.right;
    }
}

class BST {
    constructor() {
        this.root = undefined;
    }    

    /*
     * Inserts node in tree
     * Creates new node with given value and links this node after respective child or on root node
     * Values in tree must be inserted only this way, otherwise search, remove, insert would not work
     * @param {value} value to insert
     * return {void}
    */
    insert(value) {
        if (!this.root) {this.root = new Node(value, undefined); return;}

        let ptr = this.root;
        while (true) 
            if (value < ptr.value && ptr.left) ptr = ptr.left;
            else if (value >= ptr.value && ptr.right) ptr = ptr.right;
            else break;
        
        if (ptr.value > value) ptr.left = new Node(value, ptr);
        else ptr.right = new Node(value, ptr);    
    }

    /*
     * Searches node with exact value {value}
     * used by remove
     * @param {value} value to search 
     * returns pointer to node with value == given value
    */
    binarySearch(value, toAdd=0, returnLast=false) {
        if (this.root) var ptr = this.root; 
        else return;

        if (ptr.value == value) return this.root;
        while (true) {
            ptr._height += toAdd;
            if (ptr.value>value && ptr.left) ptr = ptr.left;
            else if (ptr.value<value && ptr.right) ptr = ptr.right;
            else {
                if (ptr.value == value) return ptr;
                else {
                    if (!returnLast) return undefined;
                    else return ptr;
                }
            }
        }
    }

    /*
     * Removes node with exact value
     * @param {value} value to search
    */
    /*
    * Removes node with exact value
    * @param {value} value to search
    */
    remove(value) {
        var ptr = this.binarySearch(value, -1);
        if (!ptr) return;
        
        if (ptr.left && ptr.right) {// Both childs
            let pointer = ptr.right; // check right subtree
            while (pointer.left) pointer = pointer.left; // find smallest in right subtree
            ptr.value = pointer.value; // reassign smallest value to current node
            // delete previous node with smallest value in right subtree of current node
            pointer.parent[(pointer.parent.left==pointer)?"left":"right"] = undefined;
        }
        else if (ptr.left || ptr.right) { // One child
            let atr = (ptr.right) ? "right" : "left";
            if (ptr == this.root) this.root = this.root[atr] // if ptr is root, root = root[atr]
            else {
                ptr.parent[(ptr == ptr.parent.right) ? "right" : "left"] = ptr[atr] // parent.ptr = ptr.avaivable node
                ptr[atr].parent = ptr.parent
            } 
        }
        else // Leaf Node
            ptr.parent[(ptr == ptr.parent.right) ? "right" : "left"] = undefined; // remove link to leaf node 
    }

    
    // returns height of binary tree 
    height(pointer=undefined) {
        function hght(pointer, height) {
            if (!pointer) return height;
            let left = hght(pointer.left, height+1);
            let right = hght(pointer.right, height+1);
            return (left > right ? left : right);
        }
        return hght(this.root, 0);
    }

    /*
     * Searches node with exact value {value}
     * traverse starts from top of the tree 
     * @param {value} value to search
     * returns pointer to node with value == given value
    */
    preOrderSearch(value) {
        function search(value, pointer) {
            if (pointer) {
                if (pointer.value == value) return pointer;
                let buffer = search(value, pointer.left);
                if (buffer) return buffer;
                buffer = search(value, pointer.right);
                if (buffer) return buffer;
            }
        }
        return search(value, this.root);
    }

    /*
     * Searches node with exact value {value}
     * traverse starts from very bottom of the tree 
     * @param {value} value to search
     * returns pointer to node with value == given value
    */
    postOrderSearch(value) {
        function search(value, pointer) {
            if (pointer) {
                let buffer = search(value, pointer.left);
                if (buffer) return buffer;
                buffer = search(value, pointer.right);
                if (buffer) return buffer;
                if (pointer.value == value) return pointer;
            }
        }
        return search(value, this.root);
    }

    /*
     * Searches node with exact value {value}
     * traverse starts from left bottom of the tree
     * @param {value} value to search
     * returns pointer to node with value == given value
    */
    inOrderSearch(value) {
        function search(value, pointer) {
            if (pointer) {
                let buffer = search(value, pointer.left);
                if (buffer) return buffer;
                buffer = search(value, pointer.right);
                if (buffer) return buffer;
                if (pointer.value == value) return pointer;
            }
        }
        return search(value, this.root);
    }

    /* Post Order Traverse
     * traverse from bottom of tree
     * returns string with all elements of tree 
    */
    depthFirst() { // post order traversal
        function display(ptr) {
            if (ptr) {
                let out = "";
                let buffer;
                buffer = display(ptr.left);
                if (buffer) out += buffer + " ";
                buffer = display(ptr.right);
                if (buffer) out += buffer + " "; 
                return out + ptr.value;
            }
        }
        return display(this.root);
    }

    /* In Order Traverse
     * traverse from root of tree to bottom
     * returns string with all elements of tree in tree-like look 
    */
    print() {
        function display(pointer, space, level) {
            if (!pointer) return "";
            let out=""
            space += level;
            out += display(pointer.right, space, level);
            for (let i = level; i < space; i++) out+=" ";
            out += "[" + pointer.value.toString()+ "]<\n";
            out += display(pointer.left, space, level);
            return out;
        }
        return display(this.root, 0, 5);
    }

    // returns lenght of the tree
    lenght() { // post order traversal
        function len(ptr) {
            let out = 0;
            if (ptr) {
                out += len(ptr.left);
                out += len(ptr.right);
                return out + 1;
            }
            return out;
        }
        return len(this.root);
    }
}


exports.BST = BST
exports.Node = Node
